### ISSUE TITLE
Conflict with pugixml and fmt while trying to install package

### ISSUE THREAD

**Issue by sxiii** [openmc-dev/openmc#1756]
Greetings dear OpenMC Developers!

I am using Manjaro & Archlinux and trying to install the AUR package of OpenMC from here: https://aur.archlinux.org/packages/openmc-git/

The package states "'pugixml" and "fmt" as it's conflicts. But the reality is that the only conflict it actually has is that pugixml & fmt provides some of the same files that openmc does. So if we add `--overwrite '*'` into the package installation process (to pacman -U command), it installs fine.

So, I wonder. Is there possibility to add small function to OpenMC, that checks for existence of these files, and if they exist, doesn't try to create them? (Basically, just slightly modify already-existent function). In this case, the conflicting packages of fmt and pugixml for Arch/Manjaro will no longer be a problem.

And the "solution" of uninstalling them is not always good, because a lot of other packages might be dependent on them, as I found, for example, blender.

If you need full list of files that are conflicting with these, I can provide one.

Also, this might be important as well:

![rajoo7Q](https://user-images.githubusercontent.com/575925/107072386-55874700-67e6-11eb-9317-eb38ff13225a.png)

Thanks.

**Comment by gridley** [comment]
Hey Dennis,

So, I made that build script a little bit ago, and have not put a huge amount of effort into polishing it off to make it perfect, since to my knowledge, there are three people total who have built with this method. So I'm really glad to get your feedback to improve it.

The reason for the conflicts are that we by default install pugixml and fmt versions to the system with OpenMC that are possibly of different versions than the packages you may have on your arch-based system. In fact, I didn't always have those marked as conflict, but another user of this buildscript had requested marking those as such in order to avoid actual conflicts taking place when pugixml and fmt were installed at different versions on his system, IIRC.

Ideally, we would use the system-provided fmt and pugixml packages if possible, but I've not taken the time to set some git submodules in this project as conditional, depending on whether CMake finds those on the system. This approach is potentially a drawback though, since if the system provides pugixml and fmt versions differing from those OpenMC uses by default, a successful build is not guaranteed.

If you have any suggestions on remedying this problem, I'm all ears, but do plan to leave these marked as conflicts for that reason until we can come up with a better solution.

Sorry if this has caused you any trouble,

Gavin

**Comment by sxiii** [comment]
Hi Gavin @gridley ! Thank you for your turbo-fast answer. And you never caused any troubles actually, only helped with your AUR package, which is super nice!

Yep I agree that there are "no right choice" because the versions provided by OpenMC and system might be different, but, at least, we would fix it one way or another. I see this as important task.

So ideally of course the package would ask "File conflict! Do you want to keep system file versions of the following files: ... (file list) or replace with package built-in ones? First is better for the system overall, second is better for application."... 

This is the type of question, for instance, that can be seen really often when updating a debian or ubuntu server (if they had, for example, installations of PHP and Mysql, and there was a large update). This question would solve everything for everybody (especially if we keep backup versions in case user press "Y / replace").


<details><summary>This is a code for CMAKE</summary>

```

if(EXISTS "${ROOT}/configuration/${customer}/configuration.${project_name}.xml")
   ...
else()
   ...
endif()

```

</details>

that [I googled out](https://stackoverflow.com/questions/13067401/cmake-check-that-a-local-file-exists). Some docs are also available [here](https://cmake.org/cmake/help/latest/command/if.html).

I'm no expert in Cmake and in the OpenMC project (at least yet) so I bet it's a pretty easy fix for somebody who's more into the project. :)

P.S. Also another small question: I am trying to write a PKGBUILD buildfile that will build from source the following: 1) OpenMPI 2) HDF5 3) MOAB 4) DAGMC 5) Finally, OpenMC with the support of all of the previous packages. I wonder if you know, if some of pacman/aur packages that can be used as-is without source code compilation, so the build process would be shorter & easier? Or for everything to work I should anyway compile all these from source?

Thanks a lot!

Den

**Comment by gridley** [comment]
Thanks for the notes Den @sxiii. I think it sounds like a  great plan.

On your second idea, I don't see why you'd want to build OpenMPI rather than installing the repository package for it, and same for HDF5.  As for MOAB and DAGMC, I think it would be really nice if someone put together an AUR script for that. Personally I don't use that part of OpenMC in my day to day work so have not put any thought into that.

**Comment by pshriwise** [comment]
Hi @sxiii! Thanks for taking this on. It'll be really nice to have OpenMC as an easy-to-install package on Arch Linux. 

As for MOAB and DAGMC, I don't know of any pacman/aur packages, so those will have to be built from source as well. The scripts we use to install MOAB and DAGMC in CI will likely provide some guidance here. I'd be happy to try and answer any questions you have along the way as well.

**Comment by gridley** [comment]
Alright, this should be resolved by #1757

BTW, @pshriwise, the stuff on AUR is not actually a package, but rather just scripts to automatically do the build from source. That's something I'd like to put together in the future. I think it'd be pretty easy to put it together, although I've not tried building MOAB and DAGMC from source yet.

For instance, the OpenMC build script is just:

```
# Maintainer: Gavin Ridley <gavin.keith.ridley@gmail.com>
pkgname=openmc-git
pkgver=r11702.e91a6aac1
pkgrel=1
pkgdesc="Community-developed Monte Carlo neutron and photon transport simulation code"
arch=('any')
url="https://github.com/openmc-dev/openmc"
license=('MIT')
makedepends=('cmake' 'git' 'python-setuptools')
depends=('hdf5-cpp-fortran' 'python-numpy' 'python-matplotlib' 'python-scipy'
    'python-pandas' 'python-h5py' 'python-uncertainties' 'python-lxml')
provides=("${pkgname%-git}" "libopenmc.so")
conflicts=("${pkgname%-git}" 'pugixml' 'fmt')
source=("${pkgname}::git+${url}.git")
noextract=()
sha1sums=('SKIP')
pkgver() {
	cd "${srcdir}/${pkgname}"
	printf "r%s.%s" "$(git rev-list --count HEAD)" "$(git rev-parse --short HEAD)"
}
build() {
    printf "build"
    printf "%s" "${pkgname}"
    printf "%s" "${pkgname%-git}"
	cd "${srcdir}/${pkgname}"
    git submodule update --init --recursive
    python setup.py build
    mkdir -p build
    cd build
    cmake -DCMAKE_INSTALL_PREFIX=/usr ..
    make
}
package() {
	cd "${srcdir}/${pkgname}/build"
	make DESTDIR="${pkgdir}" install
    cd ..
    python setup.py install --root="$pkgdir" --optimize=1 --skip-build
}

```

**Comment by pshriwise** [comment]
>BTW, @pshriwise, the stuff on AUR is not actually a package, but rather just scripts to automatically do the build from source. That's something I'd like to put together in the future. I think it'd be pretty easy to put it together, although I've not tried building MOAB and DAGMC from source yet. 

Ah, right. Thanks for the clarifying!
